### PR TITLE
add replay command to the CLI which point to existing replay commands

### PR DESCRIPTION
**Pull request by laura-makdah**
## Description 

Adds 3 new CLI commands for replay which call into the existing replay commands. We've removed most of the optional parameters in leu of the defaults so that we provide a simple and straightforward user experience. Based on feedback if there is an ask to expose these we can add them in as needed. 

A follow-up PR will be added to reformat the output using the tabular structure instead of the existing println! statements.

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Comment by stefan-mysten** [comment]
This looks mostly good. Are you planing for a follow up PR to handle errors too?

**Review by oxade — CHANGES_REQUESTED** [review]
Thanks for putting this up.
My biggest concern is replay currently has little to no tests as it was an internal-only debug tool. I believe it needs more polishing & tests before opening it up, lest we risk providing a bad UX for users.

File: crates/sui/src/client_commands.rs [hunk 1]
```diff
@@ -615,6 +665,54 @@ impl SuiClientCommands {
...
+                terminate_early,
+            } => {
+                let cmd = ReplayToolCommand::ReplayCheckpoints {
+                    start,
+                    end,
+                    terminate_early,
+                    max_tasks: 16,
+                };
+                // todo: automatically get non pruning fullnode url
+                let _command_result =
```
Comment (oxade) [inline crates/sui/src/client_commands.rs:711]: You don't need a non-pruning FN to replay. You just need a FN which contains the data as at the time of the TX executed.
For most users, replaying within the past few epochs is sufficient, so many FNs will do.

File: crates/sui/src/client_commands.rs [hunk 2]
```diff
@@ -1637,6 +1734,11 @@ impl Display for SuiClientCommandResult {
                 table.with(tabled::settings::style::BorderSpanCorrection);
                 writeln!(f, "{}", table)?;
             }
+            // todo: for all replay commands format results using tabular structure instead of
+            // todo: println statements in original command
```
Comment (oxade) [inline crates/sui/src/client_commands.rs:1738]: @stefan-mysten are we okay with this?
Why not polish & format first before opening it up?

File: doc/src/build/cli-client.md [hunk 3]
```diff
@@ -12,32 +12,36 @@ The Sui Client CLI installs when you install Sui. See the [Install Sui](install.
...
+| `verify-source`      | Verify local Move packages against on-chain packages, and optionally their dependencies.                                                                                                                                                      |
+| `replay-transaction` | Replay a given transaction to view transaction effects. Set environment variable MOVE_VM_STEP=1 to debug.                                                                                                                                     |
+| `replay-batch`       | Replay transactions listed in a file.                                                                                                                                                                                                         |
+| `replay-checkpoint`  | Replay all transactions in a range of checkpoints.                                                                                                                                                                                            |
```
Comment (oxade) [inline doc/src/build/cli-client.md:43]: This seems to break the existing alphabetical order

File: crates/sui/src/client_commands.rs (hunk 2 above)
Comment (stefan-mysten) [reply crates/sui/src/client_commands.rs:1738]: I am fine splitting this into multiple PRs to make it easier to review, thus my lack of commenting on this issue.

**Comment by stefan-mysten** [comment]
@oxade - most of the commands in the Sui CLI do not have proper tests, unfortunately. This is something I'd like to improve in the near future. However, if we can add more testing already to `replay` that'd be really good!

@laura-makdah regarding adding `sui replay` as a subcommand of `sui client`, I was wondering if it shouldn't be sitting in a top level command rather than a subcommand of `client`. Any pros and cons that you could think of?

@amnn - any thoughts on how to best handle this potentially large PR?

File: doc/src/build/cli-client.md (hunk 3 above)
Comment (laura-makdah) [reply doc/src/build/cli-client.md:43]: updated!

**Comment by laura-makdah** [comment]
@oxade I saw we have existing tests for replay, is there another area of coverage that you particularly would like to have? Also, I plan to work on this incrementally and will polish the format next as mentioned in the PR description :)

File: crates/sui/src/client_commands.rs [hunk 4]
```diff
@@ -607,6 +609,54 @@ pub enum SuiClientCommands {
         #[clap(long)]
         address_override: Option<ObjectID>,
     },
+
+    /// Replay a given transaction to view transaction effects. Set environment variable MOVE_VM_STEP=1 to debug.
+    #[clap(name = "replay-transaction")]
+    ReplayTransaction {
+        /// The rpc url for a non-pruning fullnode to use for fetching the transaction dependencies
+        #[arg(long = "rpc")]
+        rpc_url: String,
```
Comment (amnn) [inline crates/sui/src/client_commands.rs:618]: The CLI has a notion of an active network, which corresponds to an RPC URL -- can we rely on that as the RPC to connect to?

File: crates/sui/src/client_commands.rs [hunk 5]
```diff
@@ -607,6 +609,54 @@ pub enum SuiClientCommands {
...
+    },
+
+    /// Replay all transactions in a range of checkpoints.
+    #[command(name = "replay-checkpoint")]
+    ReplayCheckpoints {
+        /// The rpc url for a non-pruning fullnode to use for fetching the transaction dependencies
+        #[arg(long = "rpc")]
+        rpc_url: String,
+
+        /// The start value of the range of checkpoints to replay
```
Comment (amnn) [inline crates/sui/src/client_commands.rs:648]: nit: Can we be explicit about what unit/format of this argument is? I guess it's the checkpoint sequence number? I'm also not sure how pervasive that is as a unit used outside of our circles (e.g. whether people are used to checkpoints having a global order, or whether it makes sense to refer to these by the checkpoint digest).

File: doc/src/build/cli-client.md (hunk 3 above)
Comment (amnn) [reply doc/src/build/cli-client.md:43]: I don't think it was -- I still see it out of alphabetical order in`main`.

**Review by amnn — COMMENTED** [review]
Late to the party with this one, sorry!  @laura-makdah, it looks like there were some (small) unresolved comments in this PR, post-land.  From my side:

@stefan-mysten, On dealing with large PRs, I think it's fine to land incrementally and polish, as long as we can limit confusion to users until we stabilise commands.  In this case, that means I might prefer hiding these commands in `clap` (so they can still be called by us for testing, but don't show up in help output) until we are sure about names, arguments etc.

I think there was a question about `replay` being a top-level command vs a `client` command -- that could go either way.  Arguably it's a `client` command though, because it has the CLI acting as a client, talking to RPC, but to maybe question the premise of the distinction altogether, is it even useful to split commands up in that way (ones that talk to the RPC vs ones that don't?) Maybe not.

**Pull request merged**